common/compositor/gl/glprogram.cpp

```diff
@@ -128,6 +128,60 @@ static std::string GenerateFragmentShader(int layer_count) {
 
 static GLint GenerateProgram(unsigned num_textures,
                              std::ostringstream *shader_log) {
+  GLint status;
+  GLint program = glCreateProgram();
+  if (!program) {
+    if (shader_log)
+      *shader_log << "Failed to create program."
+                  << "\n";
+    return 0;
+  }
+
+  // try pre-built shader program first
+  std::ostringstream shader_program_fname;
+  shader_program_fname << "hwc_shader_prog_"
+                       << num_textures
+                       <<".shader_test.bin";
```

Comment:
Don't we need a full path here? Where would these files be stored on Android?

Reply:
Assumption was this shader program is placed in the same location as executable. I am also wondering if there's any dedicated location of those binaries loaded by compositor. I will check this and update the code.

Reply:
Do we *need* to read in the shader program from disk at runtime? It seems far safer to bake the binary data into the HWC binary.

Reply:
To make that last point more clear, you can take a look at how we are including SPIR-V (binary) shaders in the vulkan backend. https://github.com/intel/IA-Hardware-Composer/blob/4004258e14bae04ec0f6897be4f0889b4b52e0d7/common/compositor/vk/vkprogram.cpp#L22-L28

Reply:
yeah, I got your point.

Reply:
BTW, how is this binary stream generated? Also isn't this architecture independent? Does it generate  by make script?

Reply:
`glslangValidator` is used to compile the glsl down to SPIR-V format. It isn't generated during the build, we have the full binary checked into git. Now that you mention it, we really ought to have some script to automate the process for when we want to modify the shaders.

You're right, SPIR-V is architecture independent. If we don't have that luxury with pre-built GL shaders, then we need to be careful about which binaries we are generating / installing / loading.

Reply:
GL program binary is a platform dependent so we just can't have one fixed binary data as done for SPIR-V. I think this is another reason why we need to automate the process.

Reply:
I checked the fragment shader under vk/. It seems that LayerCount is fixed to 1. Does this mean only the case with 1 layer will be supported when Vulkan is used?

Reply:
SPIR-V gives you a chance to modify constants just before the program gets compiled down for the GPU. That mechanism is called specializations.
https://github.com/intel/IA-Hardware-Composer/blob/037f54610351266d0f59487821a52afc213785c8/common/compositor/vk/vkprogram.cpp#L167-L171

Reply:
I see... In case of GL (i965 DRI), the program binary is already in GPU code (with NIR attached to it in case NIR->Gen code conversion is required), so changing constants won't be allowed there. Because of this, I initially thought we should create multiple binary files each of which is from shaders with different # of layers then let HWC code to select the right shader program. We can of course get all of these binary data in the source code but the data structure might be huge if we have too many cases with different # of layers.

Reply:
Understood. So you are concerned that the sheer number of shader binaries would cause libhwcomposer to grow too large? I think the max number of layers is currently limited by number of textures the hardware supports, but I don't think we would ever hit that number on Android. Would you mind taking a look and reporting back how much memory one of these shaders actually takes? Maybe 16 shaders or so wouldn't be such a huge impact.

Reply:
Total is 579841 ~= 580KB
Yes, it is not that huge. So I assume I can probably create data structure in the source code with these binary programs.

-rw-rw-r-- 1 dongwonk dongwonk 39061 Mar 23 17:44 hwc_shader_prog_10.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 42457 Mar 23 17:44 hwc_shader_prog_11.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 45853 Mar 23 17:44 hwc_shader_prog_12.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 49233 Mar 23 17:44 hwc_shader_prog_13.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 52565 Mar 23 17:44 hwc_shader_prog_14.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 55961 Mar 23 17:44 hwc_shader_prog_15.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 59357 Mar 23 17:44 hwc_shader_prog_16.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk  8409 Mar 23 17:44 hwc_shader_prog_1.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 12069 Mar 23 17:44 hwc_shader_prog_2.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 15465 Mar 23 17:44 hwc_shader_prog_3.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 18789 Mar 23 17:44 hwc_shader_prog_4.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 22249 Mar 23 17:44 hwc_shader_prog_5.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 25573 Mar 23 17:44 hwc_shader_prog_6.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 28969 Mar 23 17:44 hwc_shader_prog_7.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 32357 Mar 23 17:44 hwc_shader_prog_8.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 35737 Mar 23 17:44 hwc_shader_prog_9.shader_test.bin
-rw-rw-r-- 1 dongwonk dongwonk 35737 Mar 23 17:44 wc_shader_prog_12.shader_test.bin

Reply:
We probably can max at 16 layers, if we do get layers beyond that fallback to compile the shader run time ?

Reply:
Also we should have README doc, if someone wants to pre-compile shaders supporting more layers if needed in their own environment.

Reply:
yes, current code already does this. If it can't find the shader program file including targeted # of layers in its name, it falls into run-time build process. By the way, I thought about getting binary data carried as a C data structure in the code but it is kinda messy unless we automate this binary generation because it's platform and i965 dependent. I am wondering if it would be acceptable to get those loaded from files now and take care of code-embedded binary later after more discussions on build automation.

Reply:
if the concern is this adds Mesa as a strong dependency here, we can take this path:
1)We can have x86 file, which has all these hard-coded c array.  Have it under build flags and enabled by default. Anyone can switch this off if not needed.
2)If implementation  is missing or 1 is disabled explicitly we can fallback loading from file system. On android the path should be /vendor/etc/ You can define it here:
https://github.com/intel/IA-Hardware-Composer/blob/master/os/android/platformdefines.h
need to come up something for Clear Linux/Yocto which can go in here: https://github.com/intel/IA-Hardware-Composer/blob/master/os/linux/platformdefines.h

Exporting the files to /vendor/etc during build time should be something we should be able to handle and @downor you can ignore it.
3)In case any of these binary shaders fail to compile or we cannot find it on system, program should fall back to run time compilation like we do now.

If you want to implement 2 first and than 1, I am k with that. But dont see much work for postponing 1 as I see you need to build all needed infrastructure for 2 anyway

Reply:
I can work on 1) however, do you want this arrays created automatically during make or you want to pre-generate these manually with fixed target and the driver version (i965_dri.so)? Having all possible arrays for different combinations is unrealistic since the shader program binary comes with sha-1 key in its header that contains i965's build id that is re-generated every time i965 is rebuilt (even if the source code is same)

Reply:
hmm k Let's go with 2 for now than while we figure out how to handle 1 @strassek Any concerns ?

Reply:
>  the shader program binary comes with sha-1 key in its header that contains i965's build id that is re-generated every time i965 is rebuilt (even if the source code is same)

So mesa rejects anything that doesn't match? This seems like a sensible restriction on Mesa's part, but I don't think we can lean on this behavior. At least not unless that restriction is coming from spec. And if it isn't, we are required to supply and load only the binaries that match the current hardware and driver build. In other words, we can't just keep pushing various binaries to Mesa until we get a match (not that anyone was suggesting otherwise).

On the HWC side how can we check if the binary we want to load is suitable for the underlying driver and hardware? Can we even ask the driver to give us its build id? @downor does shader-db give you the option of building the shader binary for anything other than the current environment?

We don't have a good reason to even attempt to shipping binaries for anything but the exact environment of the target machine, and that just isn't known until installation time.

> do you want this arrays created automatically during make or you want to pre-generate these manually with fixed target and the driver version (i965_dri.so)?

I think the reality that we are arriving at is that it just isn't tractable to pre-generate, and you can't expect products to do anything manually.

Can we instead explore the possibility of allowing the driver to generate the binaries on first run? I know that may be unacceptable for some products, but it seems like a much saner default case.

Reply:
So mesa rejects anything that doesn't match? This seems like a sensible restriction on Mesa's part, but I don't think we can lean on this behavior.

DW] This is because glProgramBinary we use to load the pre-built binary is paired with glGetProgramBinary that we also use to generate shader program binary. Both API should be from the same driver (i965) to guarantee the pre-built binary works. This is done by checking build-id embedded.

@downor does shader-db give you the option of building the shader binary for anything other than the current environment?

DW] Yes, shader-db does the cross compilation for any hardware and driver (as far as it's intel). We specify the target driver by setting "LIBGL_DRIVERS_PATH" and target architecture by passing in target GEN's PCI_ID.

Can we instead explore the possibility of allowing the driver to generate the binaries on first run? I know that may be unacceptable for some products, but it seems like a much saner default case.

DW] Unfortunately the whole point of using pre-built shaders is to make things boot up faster by letting the system bypass shader build process. BTW, "shader-cache" mechanism is already up and running in i965. This allows the system to cache shader program/object whenever it's built for the same purpose but this doesn't cover very first running.

DW] Anyway, I already started to work on a script and simple tool that generates .h with binary arrays that can be embedded in composer library. I hope we can use this for the automation. I will share these in couple of days.

Reply:
Also, if program fails to link the pre-compiled shaders we should blacklist and compile it during run time. hmm Instead of PCI_ID, can we just pass in the Gen (i.e. 7, 7.5) ?

Reply:
> We specify the target driver by setting "LIBGL_DRIVERS_PATH" and target architecture by passing in target GEN's PCI_ID.

Got it, so it sounds like we need to require that the target PCI_ID gets set at configure time, otherwise don't try enabling the feature.

Reply:
@strassek : Yeah, it should be pre-determined and yes, if not, we need to disable it.
@kalyankondapally : Each gen architecture has several variants and each variant in a group has its own PCI ID. So only way to differentiate it completely is using PCI_ID for now.